packages/app/src/heartbeatService.ts

```diff
@@ -0,0 +1,213 @@
+    const app = this.container.getProvider('app').getImmediate();
+    this._storage = new HeartbeatStorageImpl(app);
+    this._heartbeatsCachePromise = this._storage.read().then(result => {
+      this._heartbeatsCache = result;
+      return result;
+    });
+  }
+
+  /**
+   * Called to report a heartbeat. The function will generate
+   * a HeartbeatsByUserAgent object, update heartbeatsCache, and persist it
+   * to IndexedDB.
+   * Note that we only store one heartbeat per day. So if a heartbeat for today is
+   * already logged, subsequent calls to this function in the same day will be ignored.
+   */
+  async triggerHeartbeat(): Promise<void> {
+    const platformLogger = this.container
+      .getProvider('platform-logger')
+      .getImmediate();
+
+    // This is the "Firebase user agent" string from the platform logger
+    // service, not the browser user agent.
+    const userAgent = platformLogger.getPlatformInfoString();
+    const date = getUTCDateString();
+    if (this._heartbeatsCache === null) {
+      await this._heartbeatsCachePromise;
+    }
+    let heartbeatsEntry = this._heartbeatsCache!.find(
+      heartbeats => heartbeats.userAgent === userAgent
+    );
+    if (heartbeatsEntry) {
+      if (heartbeatsEntry.dates.includes(date)) {
+        // Only one per day.
+        return;
+      } else {
+        // Modify in-place in this.heartbeatsCache
+        heartbeatsEntry.dates.push(date);
+      }
+    } else {
+      // There is no entry for this Firebase user agent. Create one.
+      heartbeatsEntry = {
+        userAgent,
+        dates: [date]
+      };
+      this._heartbeatsCache!.push(heartbeatsEntry);
+    }
+    return this._storage.overwrite(this._heartbeatsCache!);
+  }
+
+  /**
+   * Returns a base64 encoded string which can be attached to the heartbeat-specific header directly.
+   * It also clears all heartbeats from memory as well as in IndexedDB.
+   *
+   * NOTE: It will read heartbeats from the heartbeatsCache, instead of from indexedDB to reduce latency
+   */
+  async getHeartbeatsHeader(): Promise<string> {
+    if (this._heartbeatsCache === null) {
+      await this._heartbeatsCachePromise;
+    }
+    // If it's still null, it's been cleared and has not been repopulated.
+    if (this._heartbeatsCache === null) {
+      return '';
+    }
+    const headerString = base64Encode(
+      JSON.stringify({ version: 2, heartbeats: this._heartbeatsCache })
+    );
+    this._heartbeatsCache = null;
+    // Do not wait for this, to reduce latency.
+    void this._storage.deleteAll();
+    return headerString;
+  }
+}
+
+function getUTCDateString(): string {
+  const today = new Date();
+  const yearString = today.getUTCFullYear().toString();
+  const month = today.getUTCMonth() + 1;
+  const monthString = month < 10 ? '0' + month : month.toString();
+  const date = today.getUTCDate();
+  const dayString = date < 10 ? '0' + date : date.toString();
+  return `${yearString}-${monthString}-${dayString}`;
```

Comment:
If you want to get rid of the custom logic here you could just do: 

```
today.toISOString().substring(0,10)
```

Reply:
Thanks! Changed.